### PR TITLE
Fix error code when exiting on OOM

### LINKED ISSUE #9979: `docker wait` can return invalid exit codes
POSIX defines process exit codes to be an unsigned 8 bit integer (0-255)
Certain cases can cause `docker wait` to return values out of that range.  For example, by intentionally invoking the OOM killer.  This can cause problems with inspecting the exit code in some shells:

```
root@ip-10-30-38-29:~# docker run -m 4m --name wat -it myjavacontainer java
root@ip-10-30-38-29:~# echo $?
255
root@ip-10-30-38-29:~# docker wait wat
-1
root@ip-10-30-38-29:~# /bin/dash -c 'exit $(docker wait wat)'
/bin/dash: 1: exit: Illegal number: -1
```
I expect `docker wait wat` to return 255, same as the exit code of `docker run`.

### PR DESCRIPTION
**Pull request by icecrime**
When exiting on OOM, the exit code simply is the (invalid) default value of -1. This patch:
- Modifies the behavior to use 137 (= [128 + SIGKILL](http://tldp.org/LDP/abs/html/exitcodes.html)).
- Adds an `oom` binary to `integration-cli/testfiles/oom` built from [this code](https://github.com/docker/docker/pull/8479#issuecomment-59556877)

Fixes #9979.

### CODE REVIEW

**Comment by SvenDowideit** [comment]
can we please document the exit codes? `Modifies the behavior to use 137 (= 128 + SIGKILL).` would let readers know we're doing this intentionally, and so if they see something  like `-1` they might report it faster :)

**Comment by tianon** [comment]
Is there some way we could please not have _another_ big binary in our Git history? :disappointed:

**Comment by LK4D4** [comment]
@tianon +1000, I think we have gcc in image.
@icecrime Nice solution, Now I wonder if there was necessary to create this ugly datastructure `ExitStatus` :) ping @vishh

**Comment by icecrime** [comment]
@tianon: Sorry for this, PTAL

@SvenDowideit: I don't know where to put this. As far as I can tell nothing in the docs actually says that `docker run` returns the container's process exit code, nor describe the OOM behavior. TBH I also believe that there's nothing specific to Docker about exit code 137 upon OOM kill.

**Comment by dqminh** [comment]
@icecrime so if the container has:
-  a parent process that spawns many child processes
- one child processes get killed by the kernel oom handler ( parent is still alive )
- a while later, parent exits on its own ( `docker stop` for example )

will the exit code be 137 or will it be normal exit code ? Disclaimer: i have not tested this yet, but the use case is probably common ( i'm looking at you bitcoin miners )

File: integration-cli/docker_cli_run_test.go [hunk 1]
```diff
@@ -2912,3 +2912,45 @@ func TestRunAllowPortRangeThroughPublish(t *testing.T) {
 	}
 	logDone("run - allow port range through --expose flag")
 }
+
+func TestRunOOMExitCode(t *testing.T) {
+	defer deleteAllContainers()
+
+	oomSource := `#include <stdlib.h>
```
Comment (LK4D4) [inline integration-cli/docker_cli_run_test.go]: nice :)

File: integration-cli/docker_cli_run_test.go [hunk 2]
```diff
@@ -2912,3 +2912,45 @@ func TestRunAllowPortRangeThroughPublish(t *testing.T) {
...
+}`
+
+	f, err := ioutil.TempFile("", "oom_")
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	f.Close()
+	gccCmd := exec.Command("gcc", "-static", "-x", "c", "-o", f.Name(), "-")
+	stdin, err := gccCmd.StdinPipe()
```
Comment (LK4D4) [inline integration-cli/docker_cli_run_test.go]: I'm not sure, but probably `gccCmd.Stdin = bytes.NewBufferString(oomSource)` can be simpler. I'm okay with this, but just FYI.

**Comment by icecrime** [comment]
@dqminh Nice catch, this is indeed the case:

```
$ docker run --rm -m 4MB -v /tmp/oom_505660737:/oom -ti busybox
/ # ./oom
Killed
/ # echo $?
137
^D
$ echo $?
137
```

My first thought would be to restrict libcontainer to notify OOM for container's PID 1 only, however as far as I can tell OOM is a cgroup wide event with no easy way to identify the killed process. WDYT @crosbymichael?

**Comment by LK4D4** [comment]
@icecrime actually we can check after oom if 1 pid is still alive

**Comment by icecrime** [comment]
@LK4D4 It will look like shit *and* probably be racy: what if PID 1 exits with code 1 when it detects one of its child died?

**Comment by LK4D4** [comment]
@icecrime You need to be fast enough :)

**Comment by dqminh** [comment]
@icecrime @LK4D4 i think in this case, the typical approach would be to check if the process received SIGKILL when it exits and return the exit code properly.. I wrote a small patch based on the current PR at https://gist.github.com/dqminh/526dfd72a60d2575cee4.

```
# file: vendor/src/github.com/docker/libcontainer/namespaces/exec.go
-       return command.ProcessState.Sys().(syscall.WaitStatus).ExitStatus(), nil
+       waitStatus := command.ProcessState.Sys().(syscall.WaitStatus)
+       exitStatus := waitStatus.ExitStatus()
+       // if the process is killed, return 137
+       signal := waitStatus.Signal()
+       if signal == syscall.SIGKILL {
+               return 128 + int(syscall.SIGKILL), nil
+       }
+       return exitStatus, nil
 }
```

So if the container's main process is killed by the kernel, it will exit with 137 status code.

**Comment by HuKeping** [comment]
@tianon @icecrime 
what about put a simple shell script in the image, and it will definitely cause OOM.

```
#! /bin/bash

i=1
x="a"

while [ True ];do
        x=$x$x
done;
```

**Comment by vishh** [comment]
@lk4d4: Since we are overloading exitCode, we can get rid of ExitStatus :)
@dqminh: I assume we will bubble up separate error codes for normal sigkills vs sigkills by OOM killer.

**Comment by tianon** [comment]
I can confirm @HuKeping's idea works:
```console
$ docker run -it --rm -m 10m debian bash
root@b8fbdfeabdbd:/# x=a; while true; do x=$x$x; done
$ echo $?
255
```

It could even be tested with a simple one-liner:
```console
$ time docker run -it --rm -m 10m debian bash -c 'x=a; while true; do x=$x$x; done'

real	0m1.770s
user	0m0.016s
sys	0m0.016s
```

**Comment by LK4D4** [comment]
@vishh Have you any idea how we can catch OOM only from init process?

**Comment by crosbymichael** [comment]
@tianon we don't have bash in our busybox container used for testing tough, that is the problem

**Comment by vishh** [comment]
@lk4d4: AFAIK there is no API to figure out the victim process of a oom
kill. Combing oom kill notification along with sigkill signal information
might be the best way forward.

**Comment by tianon** [comment]
@crosbymichael and which problem is this? :smile:
```console
$ time docker run -it --rm -m 10m busybox sh -c 'x=a; while true; do x=$x$x; done'

real	0m2.376s
user	0m0.019s
sys	0m0.017s
```

**Comment by crosbymichael** [comment]
@tianon you know I don't bash

**Comment by icecrime** [comment]
PR is updated to:

1. Rely on docker/libcontainer#328 rather than on a driver side fix
2. Keep the test case only (as fix is now in libcontainer)
2. Remove the C code from the test in favor of @HuKeping and @tianon proposal

File: integration-cli/docker_cli_run_test.go [hunk 3]
```diff
@@ -2912,3 +2912,15 @@ func TestRunAllowPortRangeThroughPublish(t *testing.T) {
 	}
 	logDone("run - allow port range through --expose flag")
 }
+
+func TestRunOOMExitCode(t *testing.T) {
+	defer deleteAllContainers()
+
+	runCmd := exec.Command(dockerBinary, "run", "-m", "4MB", "busybox", "sh", "-c", "x=a; while true; do x=$x$x; done")
```
Comment (tianon) [inline integration-cli/docker_cli_run_test.go]: This is going to be _epic_ when we have an error in our memory limit handling and this doesn't OOM. :angel:

Comment (icecrime) [inline integration-cli/docker_cli_run_test.go]: You're right, I probably should wrap the command execution in a timeout.

**Comment by icecrime** [comment]
Updated to add a timer around command execution.

**Comment by icecrime** [comment]
Rebased with #10055 merged.

**Comment by vishh** [comment]
@LK4D4, @crosbymichael: The pid that is being oom killed is logged by the kernel. It is available under '/var/log/messages'.

**Comment by jfrazelle** [comment]
LGTM

**Comment by LK4D4** [comment]
LGTM

**Pull request merged**